File: src/main/java/com/freya02/bot/docs/DocIndexMap.java
package com.freya02.bot.docs;

import com.freya02.bot.docs.index.DocIndex;
import com.freya02.botcommands.api.Logging;
import com.freya02.docs.DocSourceType;
import org.slf4j.Logger;

import java.io.IOException;
import java.util.EnumMap;

public class DocIndexMap extends EnumMap<DocSourceType, DocIndex> {
	private static final Logger LOGGER = Logging.getLogger();
	private static DocIndexMap instance;

	private DocIndexMap() {
		super(DocSourceType.class);

		Runtime.getRuntime().addShutdownHook(new Thread(() -> {
			try {
				for (DocIndex index : values()) {
					index.close();
				}
			} catch (IOException e) {
				LOGGER.error("Unable to close cache", e);
			}
		}));
	}

	public static synchronized DocIndexMap getInstance() throws IOException {
		if (instance == null) {
			instance = new DocIndexMap();

			instance.put(DocSourceType.BOT_COMMANDS, new DocIndex(DocSourceType.BOT_COMMANDS));
			instance.put(DocSourceType.JDA, new DocIndex(DocSourceType.JDA));
		}

		return instance;
	}

	public static synchronized void refreshAndInvalidateIndex(DocSourceType sourceType) throws IOException {
		final DocIndex index = instance.get(sourceType);

		if (index != null) {
			index.reindex(true);
		}
	}
}

File: src/main/kotlin/com/freya02/bot/docs/DocIndexMap.kt
package com.freya02.bot.docs

import com.freya02.bot.db.Database
import com.freya02.bot.docs.index.DocIndex
import com.freya02.docs.DocSourceType
import java.io.IOException
import java.util.*

class DocIndexMap(database: Database) : EnumMap<DocSourceType, DocIndex>(DocSourceType::class.java) {
    init {
        this[DocSourceType.BOT_COMMANDS] = DocIndex(DocSourceType.BOT_COMMANDS, database)
        this[DocSourceType.JDA] = DocIndex(DocSourceType.JDA, database)
        this[DocSourceType.JAVA] = DocIndex(DocSourceType.JAVA, database)
    }

    @Throws(IOException::class)
    suspend fun refreshAndInvalidateIndex(sourceType: DocSourceType) {
        this[sourceType]?.reindex()
    }
}